phac-nml/irida-next
# Feature: Member Access Level Update

## What does this PR do and why?
_Describe in detail what your merge request does and why._

Added in functionality to allow a member's access level to be updated from the group and project members listing page using turbo_stream. Using turbo_stream allows us to update just the access level select box for the particular member that is being updated rather than having to do a full page reload. 

A few other UI related things were added in such as an 'It's you' tag next to any member listing which matches the current user, as well as, an updated_at column to the projects and group members tables/ 

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

**Error when attempting to set a lower access level for user than they have in a parent group:**
![image](https://github.com/phac-nml/irida-next/assets/25466211/6801f565-846f-495e-bcb3-535aa1c9317c)

**Success message when an update is successful:**
![image](https://github.com/phac-nml/irida-next/assets/25466211/f7bfd5fc-170c-40ba-868b-ab6c2bc90bd0)

**It's you tag next to the member which matches the current user and the updated_at column:**
![image](https://github.com/phac-nml/irida-next/assets/25466211/2599f42f-7f3a-4ff2-be1b-b65bacfe73fd)

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

To test this out:

1) Create a group and add a few members with different access levels
2) Create a project under the above group and add members with the same roles
3) Try to update a project member's access level to any other role other than owner when they have membership in the group as an owner
4) You should be presented with an error such as in the first image under the screenshots section
5) Try updating other members access levels which don't have a higher access role in a parent group. They should update without error. To ensure that the role is actually being updated you can keep an eye on the server logs in the terminal. You should see a logging statement such as `UPDATE "members" SET "access_level" = $1, "updated_at" = $2 WHERE "members"."id" = $3`. You should also be presented with a success message as in the second image under the screenshot section

For further information on how permissions work in IRIDA Next, view the docs for  [Member Permissions](https://phac-nml.github.io/irida-next/docs/user/organization/members/member-permissions)

**Things to verify:**

1) A maintainer can update other member's access levels up to the `Maintainer` access level. Can sign in using `user6@email.com` which is only a maintainer on groups and projects to test this out.
2) A maintainer or owner cannot update their own access levels
3) A maintainer cannot update a member that has the `owner` access level
4) Setting an invalid value for a select option (using chrome -> inspect element) for one of the access levels and then updating the user to that access level should result in an error
5) A Member's access level in a project which is a part of a group can only be set to a minimum of this inherited access level
6) If a member's access level is the same in both a parent group and project, then if the member's access level in the group is updated to a higher access level, then the member's access level should also be updated to match their access level in the parent group
